### PR TITLE
Add 'Ty' formatting to `sway-fmt-v2`

### LINKED ISSUE #2101: Add `ty` formatting to `sway-fmt-v2`
Currently, if we come across a `ty` during the formatting, we simply push its string version without any formatting. This can be seen in https://github.com/FuelLabs/sway/blob/a43aec0d3fb61b7ff7875ecd0259f055cf2a32e3/sway-fmt-v2/src/items/item_enum.rs#L66-L68

I think if we have a re-usable method for formatting `ty` that would come in handy since there are a couple of more places we will need the same logic (i.e, `storage` formatting).

### Possible motivations to format a `ty`:

I am not sure how we should be formatting `ty` but if we can create a list of the motivations that might be helpful while implementing this.

1. It might not fit the total line width.
2. ?

any thoughts @eureka-cpu?

--------
## TODO

- [ ] Address #2163 
- [ ] Address #2164 
- [ ] Address #2165 
- [ ] Address #2166 
- [ ] Address #2167

### LINKED ISSUE #2163: Handle `Infer` case for `Ty` formatting
part of #2101.

There are 5 different scenarios we should consider while we are formatting a `Ty`. These are:

1. `Array`
2. `Infer`
3. `Path`
4. `Str`
5. `Tuple`

This issue is for tracking the `Infer` scenario.

### PR DESCRIPTION
**Pull request by kayagokalp**
This should be good to go for a review. One thing to note is that this PR shares the `path` formatting with #2173 (#2173 has a couple more things implemented for `path` so this `path` formatting is a subset of that one). While merging there might be some conflicts that should be easy to deal with.

## TODO
- [x] Address #2163 
- [x] Address #2164 
- [x] Address #2165 
- [x] Address #2166 
- [x] Address #2167 

closes #2101, #2163, #2164, #2165, #2166, #2167.

### CODE REVIEW

File: sway-fmt-v2/src/items/item_enum.rs
Comment (eureka-cpu) [inline sway-fmt-v2/src/items/item_enum.rs:80]: this can just be `formatted_code.push(PunctKind::Comma.as_char());`

File: sway-fmt-v2/src/utils/ty.rs
Comment (eureka-cpu) [inline sway-fmt-v2/src/utils/ty.rs:43]: Same here and a few other places, we don't need the `write` macro if we aren't joining anything I think

**Review by eureka-cpu — COMMENTED** [review]
Just a few things to fix optionally, we may also want to add something to the tests for this

File: sway-fmt-v2/src/utils/ty.rs
Comment (eureka-cpu) [inline sway-fmt-v2/src/utils/ty.rs:27]: `Path`, `Tuple` and `Array` can all implement `Format` I think

Comment (eureka-cpu) [reply sway-fmt-v2/src/utils/ty.rs:27]: But I think it could be okay either way, I'll defer to your preference

Comment (eureka-cpu) [reply sway-fmt-v2/src/utils/ty.rs:27]: If you can, I think it would be cleaner

**Review by eureka-cpu — APPROVED** [review]

File: sway-fmt-v2/src/utils/ty.rs
Comment (eureka-cpu) [reply sway-fmt-v2/src/utils/ty.rs:27]: For sure! Don't want to block you, just a suggestion to keep in mind that certain parts may benefit from using the trait instead of just duplicating the function for each one

**Comment by kayagokalp** [comment]
Decided to handle all of the `Ty` with this pr as it seems like I will not be blocked by anything. So converting this to draft until that is done.

**Comment by eureka-cpu** [comment]
Looking good so far!

File: sway-fmt-v2/src/utils/ty.rs
Comment (mitchmindtree) [reply sway-fmt-v2/src/utils/ty.rs]: One benefit of using `write!` is that it gives us the flexibility of easily refactoring the kind of buffer that we write into as `write!` supports writing into any type `T` where `T: std::fmt::Write`. I think it's fine to use `write!` generally, perhaps even preferred.

File: sway-fmt-v2/src/fmt.rs
Comment (mitchmindtree) [inline sway-fmt-v2/src/fmt.rs:278]: I realise this is just a test, but the convention for Rust and Sway types is to use capital camel case for type names, e.g. `TestTy` rather than `test_ty`. Folks sometimes refer to tests and examples upstream as demonstrative of conventions, so we should try to follow them wherever we can.

File: sway-fmt-v2/src/utils/path.rs
Comment (mitchmindtree) [inline sway-fmt-v2/src/utils/path.rs:13]: Should this be commented out? The comment mentions we can uncomment when `Ty` formatting lands, but it seems like that's happening as a part of this PR? Does this get tested yet?

**Review by mitchmindtree — CHANGES_REQUESTED** [review]
Mostly looking good! Just left a couple nits

File: sway-fmt-v2/src/fmt.rs
Comment (kayagokalp) [reply sway-fmt-v2/src/fmt.rs:278]: Ohh yeah, of course. I must have missed this one and now it triggers me pretty badly 😄

File: sway-fmt-v2/src/utils/path.rs
Comment (kayagokalp) [reply sway-fmt-v2/src/utils/path.rs:13]: I am uncommenting it but for actually testing this I think we need something like the following to the existing test which requires us to have the baseline for the `fn` formatting (which is #2173):

```rust
fn test_path_ty() { 
    match TestTy {
        TestTy::String(str) = > {},
         _ => {}
    }
}
```
I couldn't come up with another way to test that part and if we cannot find another way a test for this can be added with #2173 like above.

**Review by mitchmindtree — APPROVED** [review]
:ship:

**Review by eureka-cpu — APPROVED** [review]

**Pull request merged**
